jax/api_util.py

```diff
@@ -85,6 +86,26 @@ def apply_flat_fun_nokwargs(fun, io_tree, py_args):
   ans = fun(*args)
   return tree_unflatten(out_tree, ans)
 
+PyTreeDef = Any
+def flattened_fun_in_tree(fn: lu.WrappedFun) -> Optional[Tuple[PyTreeDef, bool]]:
+  # This implementation relies on internal details of linear_util.py's
+  # WrappedFun, but it's for the worhty cause of better user error messages.
+  # It can fail (i.e. return None) if its WrappedFun argument is not transformed
+  # with flatten_fun or flatten_fun_nokwargs, which could happen e.g. when
+  # core.eval_jaxpr encounters a call primitive (though at that point we're just
+  # round-tripping jaxprs and the user errors in question are impossible).
```

Comment:
Not that I can think of right now. It's nontrivial to plumb because it'd interact with caching (i.e. the way WrappedFun and lu.cache interact). I didn't think super hard about it though.